pkg/streamer/file.go

```diff
@@ -194,128 +195,61 @@ func fileSizeUpdated(path string, latestSize int64) (int, error) {
-	if err != nil {
-		err = terror.ErrWatcherStart.Delegate(err, dir)
-		return
-	}
-	defer watcher2.Close()
-
-	type watchResult struct {
-		updatePath string
-		err        error
-	}
-
-	result := make(chan watchResult, 1) // buffered chan to ensure not block the sender even return in the halfway
-	newCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	go func() {
-		for {
-			select {
-			case <-newCtx.Done():
-				result <- watchResult{
-					updatePath: "",
-					err:        newCtx.Err(),
-				}
+	ticker := time.NewTicker(watcherInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			if err := ctx.Err(); err != nil {
+				errCh <- err
+			}
+			return
+		case <-ticker.C:
+			// check the latest relay log file whether updated when adding watching and collecting newer
+			cmp, err := fileSizeUpdated(latestFilePath, latestFileSize)
+			switch {
+			case err != nil:
+				errCh <- terror.Annotatef(err, "latestFilePath=%s latestFileSize= %d", latestFilePath, latestFileSize)
 				return
-			case err2, ok := <-watcher2.Errors:
-				if !ok {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanClosed.Generate("errors", dir),
-					}
-				} else {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanRecvError.Delegate(err2, dir),
-					}
-				}
+			case cmp < 0:
+				errCh <- terror.ErrRelayLogFileSizeSmaller.Generate(latestFilePath)
+				return
+			case cmp > 0:
+				updatePathCh <- latestFilePath
 				return
-			case event, ok := <-watcher2.Events:
-				if !ok {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanClosed.Generate("events", dir),
-					}
+			default:
+				// current watched file size have no change means that no new writes have beed made
+				// our relay meat file will be updated immediately after receive the rotate event
+				// although we can't ensure the binlog file name in meta is the next one we expected
+				// but if we return a different filename with latestFile,the outer logic (parseDirAsPossible) will find the right one
+				metaFile, err := os.Open(filepath.Join(dir, utils.MetaFilename))
+				if err != nil {
+					errCh <- terror.Annotatef(err, "open metaFile from %s in dir %s", utils.MetaFilename, dir)
 					return
 				}
-				log.L().Debug("watcher receive event", zap.Reflect("event", event))
-				if event.IsDirEvent() {
-					log.L().Debug("skip watched event for directory", zap.Reflect("event", event))
-					continue
-				} else if !event.HasOps(watcher.Modify, watcher.Create) {
-					log.L().Debug("skip uninterested event", zap.Stringer("operation", event.Op), zap.String("path", event.Path))
-					continue
-				}
-				baseName := filepath.Base(event.Path)
-				if !binlog.VerifyFilename(baseName) {
-					log.L().Debug("skip watcher event for invalid relay log file", zap.Reflect("event", event))
-					continue // not valid binlog created, updated
+				defer metaFile.Close()
```

Comment:
I think lint should report this "defer in for-loop" 🤔 We should check why lint doesn't work first

Reply:
`Pragmatically speaking golangci-lint is a meta-linter and does not implement any linting or static analysis checks of its own. It rather relies on checks that it gets from other linters (such as govet, staticcheck, ineffassign, etc). As a result your "new check" suggestion would be more actionable if made to another GitHub project. The nature of the suggested check makes it, in my opinion, a prime candidate for go-critic`

coppied from  https://github.com/golangci/golangci-lint/issues/627

Reply:
so we should turn on which linter?

Reply:
maybe we can add a new linter ? https://github.com/go-critic/go-critic